### PR TITLE
Add PageBucketReceiver docs and remove bucketIdx TODO

### PR DESCRIPTION
**Pull request by mfussenegger**

### CODE REVIEW

File: sql/src/main/java/io/crate/jobs/PageBucketReceiver.java [hunk 1]
```diff
@@ -26,6 +26,34 @@
...
+ *
+ *  For example:
+ *
+ *   PageBucketReceiver has 2 upstream, so it expects 2 buckets.
+ *
+ *   n1: sends bucket with bucketIdx=0
+ *   n2: sends bucket with bucketIdx=1
+ *
+ *  (bucketIdx definition is up to the upstreams, but it needs to be deterministic,
+ *  see {@link io.crate.operation.projectors.DistributingDownstreamFactory#getBucketIdx(Collection)}
```
Comment (mxm) [inline sql/src/main/java/io/crate/jobs/PageBucketReceiver.java]: There is no method named `getBucketIdx`.

File: sql/src/main/java/io/crate/operation/projectors/DistributingDownstreamFactory.java [hunk 2]
```diff
@@ -112,4 +110,22 @@ public BatchConsumer create(NodeOperation nodeOperation,
...
         );
     }
+
+    /**
+     * @return bucketIdx (= idx of localNode in nodeIds)
+     *
+     * <p>
+     * This needs to be deterministic across all nodes involved in a query execution.
+     *
+     * Downstreams assume that each upstream uses a *different* bucketIdx,
```
Comment (mxm) [inline sql/src/main/java/io/crate/operation/projectors/DistributingDownstreamFactory.java]: What is used as an offset to ensure that each upstream node can generate a unique id?

File: sql/src/main/java/io/crate/operation/projectors/DistributingDownstreamFactory.java [hunk 3]
```diff
@@ -112,4 +110,22 @@ public BatchConsumer create(NodeOperation nodeOperation,
...
+     * but they don't care which node ends up with which bucketIdx.
+     * </p>
+     *
+     * See {@link io.crate.jobs.PageBucketReceiver}
+     */
+    private int getBucketIdx(Collection<String> nodeIds) {
+        ArrayList<String> server = new ArrayList<>(nodeIds);
+        server.sort(null);
+        return Math.max(server.indexOf(clusterService.localNode().getId()), 0);
```
Comment (mxm) [inline sql/src/main/java/io/crate/operation/projectors/DistributingDownstreamFactory.java]: Would it make sense to use a `SortedSet` for the node ids? And perhaps create a method `getBucketIdx(ClusterService cs)` on the `NodeOperation` which returns the idx?

Comment (mfussenegger) [reply sql/src/main/java/io/crate/operation/projectors/DistributingDownstreamFactory.java]: I'd keep it as it is - I doubt that there is a big difference between SortedSet/ArrayList in this case as the number of nodeIds is usually rather small - and with SortedSet there would be increased memory usage.

I'd also keep it here for now as it's the only place where it's used. I just wanted to improve the docs/get rid of the TODO, not change the behaviour in any other way =)

File: sql/src/main/java/io/crate/operation/projectors/DistributingDownstreamFactory.java (hunk 2 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/operation/projectors/DistributingDownstreamFactory.java]: No offset required,  nodeIds.indexOf(localNodeId) is unique. (Each node that is contained within the nodeIds of a NodeOperation, will receive that *same* NodeOperation) - Should I clarify that somehow in the comments here?

File: sql/src/main/java/io/crate/operation/projectors/DistributingDownstreamFactory.java (hunk 3 above)
Comment (mxm) [reply sql/src/main/java/io/crate/operation/projectors/DistributingDownstreamFactory.java]: Ok, just a suggestion. It is definitely an improvement as it is.

File: sql/src/main/java/io/crate/operation/projectors/DistributingDownstreamFactory.java (hunk 2 above)
Comment (mxm) [reply sql/src/main/java/io/crate/operation/projectors/DistributingDownstreamFactory.java]: Never mind, saw that the local node id is used.

**Pull request merged**